tanner/tests/test_base.py

```diff
@@ -124,3 +124,53 @@ def test_set_injectable_page(self):
             sess = session.Session(None)
         injectable_page = self.handler.set_injectable_page(sess)
         self.assertEqual(injectable_page, '/python.html')
+
+    def test_emulate_type_1(self):
+        self.handler.set_injectable_page = mock.create_autospec(self.handler.set_injectable_page)
+        data = dict(method='GET', path='/index.html',
+                    cookies={'sess_uuid': '9f82e5d0e6b64047bba996222d45e72c'})
+
+        async def mock_handle_get(session, data):
+            return {'name': 'index', 'order': 2}
+
+        self.handler.handle_get = mock_handle_get
+        detection = self.loop.run_until_complete(self.handler.emulate(data, self.session))
+        assert_detection = {'name': 'index', 'order': 2, 'type': 1, 'version': '0.6.0'}
+        self.assertEqual(detection, assert_detection)
+        self.handler.set_injectable_page.assert_not_called()
+
+    def test_emulate_type_2(self):
+        data = dict(method='GET', path='/path.html?file=/etc/passwd',
+                    cookies={'sess_uuid': '9f82e5d0e6b64047bba996222d45e72c'})
+
+        async def mock_handle_get(session, data):
+            return {'name': 'lfi', 'order': 2, 'payload': {'page': '/something.html'}}
+
+        def mock_injectable_path(session):
+            return '/path.html'
+
+        self.handler.handle_get = mock_handle_get
+        self.handler.set_injectable_page = mock_injectable_path
+
+        detection = self.loop.run_until_complete(self.handler.emulate(data, self.session))
+        assert_detection = {'name': 'lfi', 'order': 2, 'payload': {'page': '/path.html'},
+                            'type': 2, 'version': '0.6.0'}
+        self.assertEqual(detection, assert_detection)
+
+        self.handler.set_injectable_page = mock.create_autospec(self.handler.set_injectable_page)
+        self.loop.run_until_complete(self.handler.emulate(data, self.session))
```

Comment:
Why are you calling it again, just use the result from the previous one and set the mock method for the previous call.

Reply:
Yes this is the problem I was discussing, it is really hard to mock a function and simultaneously assert if it is called because `MagicMock`. Can we merge #314 first?

Reply:
Done.

Reply:
Cool!